src/Deserialization.cpp

```diff
@@ -0,0 +1,1358 @@
+        auto else_case = deserialize_stmt(if_then_else_stmt->else_case_type(), if_then_else_stmt->else_case());
+        return IfThenElse::make(condition, then_case, else_case);
+    }
+    case Serialize::Stmt_Evaluate: {
+        const Serialize::Evaluate *evaluate_stmt = (const Serialize::Evaluate *)stmt;
+        auto value = deserialize_expr(evaluate_stmt->value_type(), evaluate_stmt->value());
+        return Evaluate::make(value);
+    }
+    case Serialize::Stmt_Prefetch: {
+        const Serialize::Prefetch *prefetch_stmt = (const Serialize::Prefetch *)stmt;
+        auto name = deserialize_string(prefetch_stmt->name());
+        std::vector<Type> types;
+        types.reserve(prefetch_stmt->types()->size());
+        for (const auto &type : *prefetch_stmt->types()) {
+            types.push_back(deserialize_type(type));
+        }
+        std::vector<Range> bounds;
+        bounds.reserve(prefetch_stmt->bounds()->size());
+        for (const auto &bound : *prefetch_stmt->bounds()) {
+            bounds.push_back(deserialize_range(bound));
+        }
+        auto prefetch = deserialize_prefetch_directive(prefetch_stmt->prefetch());
+        auto condition = deserialize_expr(prefetch_stmt->condition_type(), prefetch_stmt->condition());
+        auto body = deserialize_stmt(prefetch_stmt->body_type(), prefetch_stmt->body());
+        return Prefetch::make(name, types, bounds, prefetch, condition, body);
+    }
+    case Serialize::Stmt_Acquire: {
+        const Serialize::Acquire *acquire_stmt = (const Serialize::Acquire *)stmt;
+        auto semaphore = deserialize_expr(acquire_stmt->semaphore_type(), acquire_stmt->semaphore());
+        auto count = deserialize_expr(acquire_stmt->count_type(), acquire_stmt->count());
+        auto body = deserialize_stmt(acquire_stmt->body_type(), acquire_stmt->body());
+        return Acquire::make(semaphore, count, body);
+    }
+    case Serialize::Stmt_Fork: {
+        const Serialize::Fork *fork_stmt = (const Serialize::Fork *)stmt;
+        auto first = deserialize_stmt(fork_stmt->first_type(), fork_stmt->first());
+        auto rest = deserialize_stmt(fork_stmt->rest_type(), fork_stmt->rest());
+        return Fork::make(first, rest);
+    }
+    case Serialize::Stmt_Atomic: {
+        const Serialize::Atomic *atomic_stmt = (const Serialize::Atomic *)stmt;
+        auto producer_name = deserialize_string(atomic_stmt->producer_name());
+        auto mutex_name = deserialize_string(atomic_stmt->mutex_name());
+        auto body = deserialize_stmt(atomic_stmt->body_type(), atomic_stmt->body());
+        return Atomic::make(producer_name, mutex_name, body);
+    }
+    case Serialize::Stmt_UndefinedStmt: {
+        return Stmt();
+    }
+    default:
+        user_assert(false) << "unknown type code " << type_code << "\n";
+        return Stmt();
+    }
+}
+
+Expr Deserializer::deserialize_expr(Serialize::Expr type_code, const void *expr) {
+    assert(expr != nullptr);
```

Comment:
...wait, is `Serialize::Expr` a pointer type?

Reply:
no, it's an enum type underneath the hood